KAFKA-5664: Disable auto offset commit in ConsoleConsumer if no group is provided

vahidhashemian opened the pull request:
This is to avoid polluting the Consumer Coordinator cache as the auto-generated group and its offsets are unlikely to be reused.

vahidhashemian committed:
KAFKA-5664: Disable auto offset commit in ConsoleConsumer if no group is provided

This is to avoid polluting the Consumer Coordinator cache as the group and its offsets are unlikely to be reused.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6418/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6403/
Test FAILed (JDK 8 and Scala 2.12).

hachikuji commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -405,6 +405,8 @@ object ConsoleConsumer extends Logging {
     if (!consumerProps.containsKey(ConsumerConfig.GROUP_ID_CONFIG)) {
       consumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, s"console-consumer-${new Random().nextInt(100000)}")
+      // Avoid polluting the coordinator cache since the group won't likely be used again
+      consumerProps.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false")
```
Should we override only if the config is not present?

vahidhashemian replied:
Yes, that makes sense. I'll update the patch. Thanks.
Does it need to be documented somewhere that only for the console consumer the default for `enable.auto.commit` is `false`?

hachikuji replied:
Wouldn't hurt to mention it in the release notes.

vahidhashemian replied:
Thanks. I just updated the PR.

vahidhashemian committed:
Added a check for when the property is explicitly provided

Also added a release note for how the default property value is changed for the particular scenario.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6412/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6427/
Test PASSed (JDK 7 and Scala 2.11).

hachikuji commented:
So there is one case I can think of where this could cause an observable change in behavior. If the group is dynamically generated, I think it's safe to assume that we have only one member. But even with a group of 1, it is possible for a rebalance to get triggered. For example, the consumer could lose connectivity with the broker and cause a session timeout. In that case, after rebalancing, we will reset the offset instead of using the last committed offset (since there will be none). I am not sure if this is a major issue. Maybe at least we should log a warning if the group rebalances and the offset gets reset? Thoughts?

cc @ijuma

vahidhashemian commented:
The scenario you described should not be a common one. But it makes sense to somehow notify the user of the behavior change. Do you think the warning should be logged through a custom rebalance listener?

hachikuji commented:
@vahidhashemian I think if we're going to the trouble to add a custom rebalance listener, then we may as well implement local offset caching to preserve the current behavior. 

Thinking about this issue a little more, we don't actually have any reports of problems resulting from the extra state from these groups. The reason I created the JIRA is that we were considering increasing the default offset expiration time which would tend to increase cache waste. But perhaps until we do so, we should err on the side of compatibility? Also, if we fix KAFKA-4682, the argument to increase the expiration time might be a little weaker.

vahidhashemian commented:
@hachikuji What I understand from your comment is that if we fix KAFKA-4682 we may not need to do this PR. It sounds reasonable to me, and we can hold off. Please advise if I misunderstood your point. Thanks.

vahidhashemian commented:
Submitted as part of [this PR](https://github.com/apache/kafka/pull/4896).
